subs/curator/src/Curator/Snapshot.hs

```diff
@@ -92,3 +112,287 @@ traverseValidate f t = do
+      Just (Just cabalVersion) -> do
+        pkgInfos <- Map.traverseWithKey (getPkgInfo constraints) (snapshotPackages snapshot)
+        let depTree =
+              Map.map (piVersion &&& piTreeDeps) pkgInfos
+              <> Map.map (\mv -> (mv, [])) ghcBootPackages
+        return $ Map.mapWithKey (validateDeps constraints depTree cabalVersion) pkgInfos
+    let (rangeErrors, otherErrors) = splitErrors pkgErrors
+    unless (Map.null rangeErrors && Map.null otherErrors) $ do
+      logWarn "Errors in snapshot:"
+      void $ flip Map.traverseWithKey rangeErrors $ \(dep, maintainers, mver) users -> do
+        logWarn $ RIO.display (pkgBoundsError dep maintainers mver users)
+      void $ flip Map.traverseWithKey otherErrors $ \pname errors -> do
+        logWarn $ fromString (packageNameString pname)
+        forM_ errors $ \err -> logWarn $ "    " <> fromString err
+
+pkgBoundsError ::
+       PackageName
+    -> Set Text
+    -> (Maybe Version)
+    -> Map DependingPackage DepBounds
+    -> Text
+pkgBoundsError dep maintainers mdepVer users =
+    T.unlines $ ""
+              : showDepVer
+              : map showUser (Map.toList users)
+  where
+    showDepVer | Just version <- mdepVer =
+                   T.concat [ display dep , "-" , display version
+                            , displayMaintainers maintainers
+                            , " is out of bounds for:"
+                            ]
+               | otherwise =
+                   T.concat [ display dep, displayMaintainers maintainers
+                            , " (not present) depended on by:"
+                            ]
+
+    displayMaintainers ms | Set.null ms = ""
+                          | otherwise = T.concat [" (", T.intercalate ", " (Set.toList ms), ")"]
+
+    showUser :: (DependingPackage, DepBounds) -> Text
+    showUser (dp, db) = T.concat
+            [ "- [ ] "
+            , depPackageShow1 dp
+            , " ("
+            -- add a space after < to avoid confusing Markdown processors (like
+            -- Github's issue tracker)
+            , TP.replace "<" "< " $ display (dbRange db)
+            , "). "
+            , depPackageShow2 dp
+            , ". Used by: "
+            , T.intercalate ", " $ map compToText $ Set.toList (dbComponents db)
+            ]
+
+    depPackageShow1 :: DependingPackage -> Text
+    depPackageShow1 DependingPackage {..}
+      | Just v <- dpVersion = display dpName <> "-" <> display v
+      | otherwise = display dpName
+
+    depPackageShow2 :: DependingPackage -> Text
+    depPackageShow2 DependingPackage {..} = T.intercalate ". " $
+        ( if null dpMaintainers
+          then "No maintainer"
+          else T.intercalate ", " (Set.toList dpMaintainers)
+        ) :
+        if null dpGithubPings
+        then []
+        else [T.concat (map (T.cons '@') (Set.toList dpGithubPings))]
+
+    compToText :: Component -> Text
+    compToText CompLibrary = "library"
```

Comment:
We can defer this question to after this PR, but: how do we handle dependencies of custom-setup stanzas, foreign libraries, and sublibraries? Do they all appear as library dependencies?

Reply:
I had a thought about those but somehow forgot about it. This code doesn't handle them and I think it makes sense to fix it as a part of this PR